src/gwt/src/org/rstudio/studio/client/dataviewer/DataTable.java

```diff
@@ -98,7 +97,98 @@ public void onValueChange(ValueChangeEvent<String> event)
          toolbar.addRightWidget(label);
       }
    }
-   
+
+   private ClickHandler[] getColumnViewClickHandlers() {
+      ClickHandler handlers[] = {
+         new ClickHandler()
+         {
+            @Override
+            public void onClick(ClickEvent event)
+            {
+               firstColumnPage();
+            }
+         },
+         new ClickHandler()
+         {
+            @Override
+            public void onClick(ClickEvent event)
+            {
+               prevColumnPage();
+            }
+         },
+         new ClickHandler()
+         {
+            @Override
+            public void onClick(ClickEvent event)
+            {
+               nextColumnPage();
+            }
+         },
+         new ClickHandler()
+         {
+            @Override
+            public void onClick(ClickEvent event)
+            {
+               lastColumnPage();
+            }
+         }
+      };
+
+      return handlers;
+   }
+   private void addColumnControls(Toolbar toolbar)
+   {
+      colsLabel_ =
+         new ToolbarLabel("Cols:");
+      colsLabel_.addStyleName(ThemeStyles.INSTANCE.toolbarInfoLabel());
+      colsLabel_.setVisible(false);
+      toolbar.addLeftWidget(colsLabel_);
+
+      ClickHandler[] clickHandlers = getColumnViewClickHandlers();
+      SimpleButton columnButton;
+      columnViewButtons_ = new ArrayList<>();
+
+      for (int i = 0; i < COLUMN_VIEW_BUTTONS.length; i++)
+      {
+         columnButton = new SimpleButton(COLUMN_VIEW_BUTTONS[i]);
+         columnButton.addClickHandler(clickHandlers[i]);
+         columnButton.setVisible(false);
+         toolbar.addLeftWidget(columnButton);
+         columnViewButtons_.add(columnButton);
+
+         if (i == 1)
+         {
+            columnTextWidget_ = new DataTableColumnWidget(this::setOffsetAndMaxColumns);
+            columnTextWidget_.setVisible(false);
+            toolbar.addLeftWidget(columnTextWidget_);
+         }
+      }
+   }
+
+   private void setColumnControlVisibility(boolean visible)
+   {
+      colsLabel_.setVisible(false);
+      for (int i = 0; i < COLUMN_VIEW_BUTTONS.length; i++)
+      {
+         columnViewButtons_.get(i).setVisible(visible);
+      }
+      columnTextWidget_.setVisible(visible);
+   }
+
+   private CommandWith2Args<Integer, Integer> getDataTableColumnCallback()
+   {
+      return (offset, max) ->
+      {
+         // This is super weird but the type coming in is throwing crazy exceptions if
```

Comment:
This might be because GWT's integer type is a wrapped integer as opposed to a 'native' JavaScript number.

Reply:
Ah interesting! I tried doing all sorts of debug logging to try to get to the meat of the issue but this was the only solution that ended up working for me. Would love to try something cleaner.